File: src/bin/0227-basic-calculator-ii.py
# @leetup=info id=227 lang=python3 slug=basic-calculator-ii



from typing import Optional

# @leetup=code
class Solution:
    def calculate(self, s: str) -> int:
        # Append "+" as sentinel.
        s = s + "+"
        nums = []
        num = 0
        op = "+"
        for c in s:
            if c.isdigit():
                num = num * 10 + ord(c) - ord("0")
            elif c != " ":
                if op == "+":
                    nums.append(num)
                elif op == "-":
                    # 22 - 3 * 5
                    #        ^
                    # nums = [22]
                    # op = '-'
                    # num = 3
                    nums.append(-num)
                    # nums = [22, -3]
                    # op = '*' (later)
                elif op == "*":
                    # 22 - 3 * 5 + 1
                    #            ^
                    # nums = [22, 3]
                    # op = '*'
                    # num = 5

                    nums[-1] *= num
                    # nums = [22, 15]
                    # op = '+'
                    # num = 0
                elif op == "/":
                    nums[-1] = int(float(nums[-1]) / num)

                op = c
                num = 0
        return sum(nums)


# @leetup=code

File: src/bin/0227-basic-calculator-ii.rs
// @leetup=info id=227 lang=rust slug=basic-calculator-ii

// @leetup=code

#[allow(unused_imports)]
use std::collections::*;

impl Solution {
    pub fn calculate(mut s: String) -> i32 {
        let mut nums = Vec::new();
        // Append "+" as sentinel
        s.push_str("+");
        let mut num = 0;
        // previous operation
        let mut op = '+';
        for c in s.chars() {
            if let Some(d) = c.to_digit(10) {
                num = num * 10 + (d as i32)
            } else if !c.is_whitespace() {
                match op {
                    '+' => nums.push(num),
                    '-' => nums.push(-num),
                    '*' => *nums.last_mut().unwrap() *= num,
                    '/' => *nums.last_mut().unwrap() /= num,
                    _ => unreachable!()
                }
                op = c;
                num = 0;
            }
        }
        nums.iter().sum()
    }
}
// @leetup=code

struct Solution;

fn main() {
    let _ = Solution::calculate;
}
